src/dune_lang/parser.ml

```diff
@@ -13,16 +13,73 @@ let error (loc : Loc.t) message = User_error.raise ~loc [ Pp.text message ]
 
    We could also do clever things with GADTs, but it will add type variables
    everywhere which is annoying. *)
-let rec cst_of_encoded_ast (x : Ast.t) : Cst.t =
-  match x with
-  | Template t -> Template t
-  | Quoted_string (loc, s) -> Quoted_string (loc, s)
-  | List (loc, l) -> List (loc, List.map l ~f:cst_of_encoded_ast)
-  | Atom (loc, (A s as atom)) -> (
-    match s.[0] with
-    | '\000' -> Comment (loc, Lines (String.drop s 1 |> String.split ~on:'\n'))
-    | '\001' -> Comment (loc, Legacy)
-    | _ -> Atom (loc, atom))
+module Encoded : sig
+  type t
+
+  val template : Template.t -> t
+
+  val atom : Loc.t -> Atom.t -> t
+
+  val quoted_string : Loc.t -> string -> t
+
+  val comment : Loc.t -> string list -> t
+
+  val list : Loc.t -> t list -> t
+
+  val to_csts : t list -> Cst.t list
+
+  val to_asts : t list -> Ast.t list
+end = struct
+  open Ast
+
+  type t = Ast.t
+
+  let comment_marker : Template.t =
```

Comment:
Ugh, I wonder if we could get rid of these nonsensical markers using a variant type.